src/serialize-closure.ts

```diff
@@ -0,0 +1,987 @@
+      if (funcAST === undefined) {
+        throw new Error(`SourceFile did not contain a FunctionDeclaration`);
+      }
+
+      if (serializeProps.preProcess) {
+        // allow callers to pre-process the AST before serialization
+        funcAST = transform(funcAST, serializeProps.preProcess);
+      }
+
+      // walk the AST and resolve any free variables - i.e. any ts.Identifiers that point
+      // to a value outside of the closure's scope.
+      const freeVariables = (
+        await Promise.all(discoverFreeVariables(funcAST))
+      ).filter((a): a is Exclude<typeof a, undefined> => a !== undefined);
+
+      // when naming variables,
+      const illegalNames = new Set([
+        ...freeVariables.map(({ variableName }) => variableName),
+        // collect all ids used within the closure
+        // we will use this to ensure there are no name conflicts with hoisted closure variables
+        ...discoverFunctionIdentifiers(funcAST!),
+      ]);
+
+      // hoist free variable declarations
+      for (const freeVariable of freeVariables) {
+        const serializedFreeVariable = await serializeValue(freeVariable, {
+          illegalNames,
+        });
+      }
+
+      if (serializeProps.postProcess?.length) {
+        // allow user to apply AST post-processing to the serialized closure
+        funcAST = transform(funcAST, serializeProps.postProcess);
+      }
+
+      // emit a statement to the bundle that instantiates this closure
+      emit(
+        [
+          await createClosureStatement(
+            closureName,
+            parsedCode.statements[0],
+            freeVariables
+          ),
+        ],
+        parsedCode
+      );
+
+      // return an Identifier pointing to the serialized and initialized closure
+      return ts.factory.createIdentifier(closureName);
+
+      /**
+       * Serializes a JavaScript value into the bundle.
+       *
+       * See https://twitter.com/samgoodwin89/status/1545618799482642433?s=20&t=AxGG-nfqH_KQHzmkEBKZ8g
+       * for a diagram explaining the algorithmic flow.
+       *
+       * ## General Form
+       *
+       * A Function is emitted in the following general form:
+       * ```ts
+       * export const f1 = ((...freeVariables) => function Foo() {
+       *   // syntax
+       * })(_self, ...freeVariables)
+       * f1.prototype = (serialized Foo.prototype)
+       * // (add static properties)
+       * f1.staticProp == staticValue
+       * // .. etc.
+       * // (add static prototype)
+       * Object.setPrototype(f1, Function);
+       *
+       * ```
+       *
+       * Objects are emitted as a literal, followed by a bunch of statements setting values on the object.
+       * ```ts
+       * export const v1 = Object.create(BaseType.prototype);
+       * v1.constructor = f1;
+       * v1.prop = value1;
+       * v2.prop = value2;
+       * // ..
+       * vn.prop = valueN;
+       * ```
+       *
+       * ## Algorithm
```

Comment:
Still working out the exact algorithm, but making great progress thanks to this visualization.
![image](https://user-images.githubusercontent.com/38672686/178099858-95040271-ded7-4820-8601-942b26f80c22.png)